cmd/kubelet/app/server.go

```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 	return nil
 }
 
+// buildKubeletClientConfig constructs the appropriate client config for the kubelet depending on whether
+// bootstrapping is enabled or client certificate rotation is enabled.
+func buildKubeletClientConfig(s *options.KubeletServer, nodeName types.NodeName) (*restclient.Config, func(), error) {
+	if s.RotateCertificates && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletClientCertificate) {
+		certConfig, clientConfig, err := bootstrap.LoadClientConfig(s.KubeConfig, s.BootstrapKubeconfig, s.CertDirectory)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newClientFn := func(current *tls.Certificate) (certificatesclient.CertificateSigningRequestInterface, error) {
+			// If we have a valid certificate, use that to fetch CSRs. Otherwise use the bootstrap
+			// credentials.
+			// XXX: When an external bootstrap source is available, it should be possible to always use that source
+			//      to retrieve new credentials.
+			config := certConfig
+			if current != nil {
+				config = clientConfig
+			}
+			client, err := clientset.NewForConfig(config)
+			if err != nil {
+				return nil, err
+			}
+			return client.CertificatesV1beta1().CertificateSigningRequests(), nil
+		}
+
+		clientCertificateManager, err := kubeletcertificate.NewKubeletClientCertificateManager(
+			s.CertDirectory,
+			nodeName,
+			clientConfig.CertFile,
+			clientConfig.KeyFile,
+			newClientFn,
+		)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// the rotating transport will use the cert from the cert manager instead of these files
+		transportConfig := *clientConfig
+		transportConfig.CertFile = ""
+		transportConfig.KeyFile = ""
+
+		// we set exitAfter to five minutes because we use this client configuration to request new certs - if we are unable
+		// to request new certs, we will be unable to continue normal operation. Exiting the process allows a wrapper
+		// or the bootstrapping credentials to potentially lay down new initial config.
+		closeAllConns, err := kubeletcertificate.UpdateTransport(wait.NeverStop, &transportConfig, clientCertificateManager, 5*time.Minute)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		glog.V(2).Info("Starting client certificate rotation.")
+		clientCertificateManager.Start()
+
+		return &transportConfig, closeAllConns, nil
+	}
+
+	if len(s.BootstrapKubeconfig) > 0 {
```

Comment:
Should this get called before starting the cert managed above?

Reply:
this is in the unmanaged path.  In the managed path bootstrap is managed by the manager.